_posts/2023-09-18-aws-cognito-azure-ad.md

```diff
@@ -0,0 +1,262 @@
+  domain       = "example-application"
+  user_pool_id = aws_cognito_user_pool.cognito_user_pool.id
+}
+```
+
+{:refdef: style="text-align: center;"}
+<img src="{{ '/img/2023-09-18-cognito-azure-ad/cognito-domain.jpg' | prepend: site.baseurl }}" alt="AWS Cognito: Domain" class="image fit" style="margin:0px auto; max-width:100%">
+_Domain_
+{: refdef}
+
+To utilize users from our Azure AD, we must integrate an identity provider rather than generating new users within Cognito.
+In this integration, we'll reference the previously created client_id and client_secret from Azure AD.
+
+Additionally, as we aim to synchronize all useful user attributes from Azure AD to AWS Cognito, we're required to complete an attribute mapping. 
+The data in the response body of the `attributes_url` variable will be aligned with the corresponding properties in AWS Cognito.
+
+**Please note: Don't forget to fill in your tenant ID.**
+
+```terraform
+resource "aws_cognito_identity_provider" "cognito_identity_provider" {
+  user_pool_id  = aws_cognito_user_pool.cognito_user_pool.id
+  provider_name = "AzureAD"
+  provider_type = "OIDC"
+
+  provider_details = {
+    authorize_scopes          = "email profile openid"
+    client_id                 = azuread_application.application.application_id
+    client_secret             = azuread_application_password.application_password.value
+    oidc_issuer               = "https://login.microsoftonline.com/${var.tenant_id}/v2.0"
+    authorize_url             = "https://login.microsoftonline.com/${var.tenant_id}/oauth2/v2.0/authorize"
+    token_url                 = "https://login.microsoftonline.com/${var.tenant_id}/oauth2/v2.0/token"
+    attributes_request_method = "GET"
+    attributes_url            = "https://graph.microsoft.com/oidc/userinfo"
+    jwks_uri                  = "https://login.microsoftonline.com/${var.tenant_id}/discovery/v2.0/keys"
+  }
+
+  attribute_mapping = {
+    email          = "email"
+    name           = "name"
+    given_name     = "given_name"
+    family_name    = "family_name"
+    picture        = "picture"
+    email_verified = "email_verified"
+  }
+}
+```
+{:refdef: style="text-align: center;"}
+<img src="{{ '/img/2023-09-18-cognito-azure-ad/cognito-azuread-identity-provider.jpg' | prepend: site.baseurl }}" alt="AWS Cognito: Domain" class="image fit" style="margin:0px auto; max-width:100%">
+_Azure AD Identity Provider_
+{: refdef}
+
+For the integration of the authentication process within services such as API Gateway, having the user pool ID is necessary.
+This is why we are storing the ID in the Parameter Store, enabling us to retrieve it whenever necessary.
+```terraform
+resource "aws_ssm_parameter" "ssm_parameter" {
+  name  = "example-application-for-aws-user-pool-id"
+  type  = "String"
+  value = aws_cognito_user_pool.cognito_user_pool.id
+}
+```
+# Conclusion
+By following this approach, we can seamlessly authenticate our enterprise users across AWS services.
```

Comment:
In conclusion explain what this has solved compared to alternatives. Integration with other AWS services is a key factor, explain this in more detail. 

Explain what can be improved on the example to make it production ready, if anything. 

Maybe as a final parting thought (optional) add that this is only for allowing applications to authenticate against cognito, it's not to setup SSO for AWS itself.